File: production_api/production_api/report/work_order_report/work_order_report.js
// Copyright (c) 2025, Essdee and contributors
// For license information, please see license.txt

function toggle_based_on_filters() {
	const based_on = frappe.query_report.get_filter_value("based_on") || "Date";
	const from_date_filter = frappe.query_report.get_filter("from_date");
	const to_date_filter = frappe.query_report.get_filter("to_date");
	const status_filter = frappe.query_report.get_filter("status");

	const show_date = based_on === "Date";
	const show_status = based_on === "WO Status";

	[from_date_filter, to_date_filter].forEach((f) => {
		if (!f) return;
		f.df.hidden = show_date ? 0 : 1;
		f.df.reqd = show_date ? 1 : 0;
		f.refresh();
	});

	if (status_filter) {
		status_filter.df.hidden = show_status ? 0 : 1;
		status_filter.df.reqd = 0;
		status_filter.refresh();
	}
}

frappe.query_reports["Work Order Report"] = {
	"onload": function () {
		toggle_based_on_filters();
	},
	"filters": [
		{
			"fieldname": "based_on",
			"fieldtype": "Select",
			"label": "Based On",
			"options": "Date\nWO Status",
			"default": "Date",
			"reqd": 1,
			"on_change": function () {
				toggle_based_on_filters();
				frappe.query_report.refresh();
			},
		},
		{
			"fieldname":"from_date",
			"fieldtype":"Date",
			"label":"From Date",
			"default": frappe.datetime.add_months(frappe.datetime.nowdate(), -1),
			"reqd":1,
		},
		{
			"fieldname":"to_date",
			"fieldtype":"Date",
			"label":"To Date",
			"default":frappe.datetime.nowdate(),
			"reqd":1,
		},
		{
			"fieldname":"supplier",
			"fieldtype":"Link",
			"options":"Supplier",
			"label":"Supplier",
		},
		{
			"fieldname":"process_name",
			"fieldtype":"Link",
			"options":"Process",
			"label":"Process",
		},
		{
			"fieldname":"item",
			"fieldtype":"Link",
			"options":"Item",
			"label":"Item",
		},
		{
			"fieldname":"lot",
			"fieldtype":"Link",
			"options":"Lot",
			"label":"Lot",
		},
		{
			"fieldname":"status",
			"fieldtype":"Select",
			"options":"\nOpen\nClose",
			"label":"WO Status",
		}
	]
};

File: production_api/production_api/report/work_order_report/work_order_report.py
# Copyright (c) 2025, Essdee and contributors
# For license information, please see license.txt

import frappe, json
from six import string_types

def execute(filters=None):
	types = frappe.db.get_single_value("MRP Settings", "received_type_list")
	received_types = types.split(",")
	for t in received_types:
		t = t.strip()
	columns = get_columns(received_types)
	data = get_data(filters, received_types)
	return columns, data

def get_columns(received_types):
	columns = [
		{"fieldname":"work_order","fieldtype":"Link","options":"Work Order","label":"Work Order"},
		{"fieldname":"wo_date","fieldtype":"Date","label":"WO Date"},
		{"fieldname":"supplier","fieldtype":"Link","options":"Supplier","label":"Supplier"},
		{"fieldname":"supplier_name","fieldtype":"Data","label":"Supplier Name"},
		{"fieldname":"lot","fieldtype":"Link","options":"Lot","label":"Lot"},
		{"fieldname":"item","fieldtype":"Link","options":"Item","label":"Item"},
		{"fieldname":"wo_colours","fieldtype":"Data","label":"WO Colours"},
		{"fieldname":"process_name","fieldtype":"Link","options":"Process","label":"Process"},
		{"fieldname":"planned_quantity","fieldtype":"Int","label":"Pieces Planned"},
		{"fieldname":"total_no_of_pieces_delivered","fieldtype":"Int","label":"Pieces Delivered"},
		{"fieldname":"total_no_of_pieces_received","fieldtype":"Int","label":"Pieces Received"},
	]

	for t in received_types:
		x = t.replace(" ", "_")
		columns.append(
			{"fieldname": x,"fieldtype":"Int","label":t},
		)
	columns = columns + [
		{"fieldname":"others", "fieldtype":"Int","label":"Others"},
		{"fieldname":"pending","fieldtype":"Int","label":"Pending"},	
		{"fieldname":"first_dc_date","fieldtype":"Date","label":"First DC Date"},
		{"fieldname":"last_dc_date","fieldtype":"Date","label":"Last DC Date"},
		{"fieldname":"first_grn_date","fieldtype":"Date","label":"First GRN Date"},
		{"fieldname":"last_grn_date","fieldtype":"Date","label":"Last GRN Date"},
	]	
	return columns

def get_data(filters, received_types):
	doctype = frappe.qb.DocType("Work Order")
	query = frappe.qb.from_(doctype).select(
		doctype.name.as_("work_order"),
		doctype.wo_date,
		doctype.supplier,
		doctype.supplier_name,
		doctype.lot,
		doctype.item,
		doctype.wo_colours,
		doctype.process_name,
		doctype.planned_quantity,
		doctype.total_no_of_pieces_delivered,
		doctype.total_no_of_pieces_received,
		(doctype.total_no_of_pieces_received - doctype.total_no_of_pieces_delivered).as_("pending"),
		doctype.first_dc_date,
		doctype.last_dc_date,
		doctype.first_grn_date,
		doctype.last_grn_date,
		doctype.received_types_json
	).where((doctype.docstatus == 1) & (doctype.wo_date >= filters.get("from_date")) & (doctype.wo_date <= filters.get("to_date")))


	if supplier := filters.get('supplier'):
		query = query.where(doctype.supplier == supplier)
	if process := filters.get("process_name"):
		query = query.where(doctype.process_name == process)
	if lot := filters.get("lot"):
		query = query.where(doctype.lot == lot)
	if item := filters.get("item"):
		query = query.where(doctype.item == item)	
	if status := filters.get("status"):
		if status == "Close":
			query = query.where(doctype.open_status == "Close")
		elif status == "Open":
			query = query.where(doctype.open_status == "Open")

	result = query.run(as_dict=True)
	for res in result:
		received_json = res['received_types_json']
		if isinstance(received_json, string_types):
			received_json = json.loads(received_json)
		others = 0	
		if received_json:
			for t in received_json:
				if t not in received_types:
					others += received_json.get(t)
				elif t in received_types and received_json.get(t):
					x = t.replace(" ", "_")
					res.update({x: received_json.get(t)})
				else:
					x = t.replace(" ", "_")
					res.update({x: 0})

		res.update({"others":others})	

	return result
